packages/firestore/test/integration/api/get_options.test.ts

```diff
@@ -0,0 +1,612 @@
+      });
+    });
+  });
+
+  it('get document while offline with default get options', () => {
+    const initialData = { key: 'value' };
+    const newData = { key2: 'value2' };
+    return withTestDoc(persistence, docRef => {
+      return docRef
+        .set(initialData)
+        .then(() => docRef.firestore.disableNetwork())
+        .then(() => {
+          // NB: since we're offline, the returned promise won't complete
+          docRef.set(newData);
+          return docRef.get();
+        })
+        .then(doc => {
+          expect(doc.exists).to.be.true;
+          expect(doc.metadata.fromCache).to.be.true;
+          expect(doc.metadata.hasPendingWrites).to.be.true;
+          expect(doc.data()).to.deep.equal(newData);
+        });
+    });
+  });
+
+  it('get collection while offline with default get options', () => {
+    const initialDocs = {
+      doc1: { key1: 'value1' },
+      doc2: { key2: 'value2' },
+      doc3: { key3: 'value3' }
+    };
+    return withTestCollection(persistence, initialDocs, colRef => {
+      // force local cache of these
+      return (
+        colRef
+          .get()
+          // now go offine. Note that if persistence is disabled, this will cause
+          // the initialDocs to be garbage collected.
+          .then(ignored => colRef.firestore.disableNetwork())
+          .then(() => {
+            // NB: since we're offline, the returned promises won't complete
+            colRef.doc('doc2').set({ key2b: 'value2b' }, { merge: true });
+            colRef.doc('doc3').set({ key3b: 'value3b' });
+            colRef.doc('doc4').set({ key4: 'value4' });
+            return colRef.get();
+          })
+          .then(qrySnap => {
+            expect(qrySnap.metadata.fromCache).to.be.true;
+            expect(qrySnap.metadata.hasPendingWrites).to.be.true;
+            let docsData = {};
+            qrySnap.forEach(doc => {
+              docsData[doc.id] = doc.data();
+            });
+            if (persistence) {
+              expect(qrySnap.docChanges.length).to.equal(4);
+              expect(docsData).to.deep.equal({
+                doc1: { key1: 'value1' },
+                doc2: { key2: 'value2', key2b: 'value2b' },
+                doc3: { key3b: 'value3b' },
+                doc4: { key4: 'value4' }
+              });
+            } else {
+              expect(qrySnap.docChanges.length).to.equal(3);
+              expect(docsData).to.deep.equal({
+                doc2: { key2b: 'value2b' },
+                doc3: { key3b: 'value3b' },
+                doc4: { key4: 'value4' }
+              });
+            }
+          })
+      );
+    });
+  });
+
+  it('get document while online cache only', () => {
+    const initialData = { key: 'value' };
+    return withTestDoc(persistence, docRef => {
+      return docRef
+        .set(initialData)
+        .then(() => docRef.get({ source: 'cache' }))
+        .then(doc => {
+          if (persistence) {
```

Comment:
As mentioned earlier, you could register a listener to control whether the document is in the cache, rather than having conditional test logic.